numpy/linalg/umath_linalg.c.src

```diff
@@ -3587,6 +3617,15 @@ GUFUNC_DESCRIPTOR_t gufunc_descriptors [] = {
         4, 3, 4,
         FUNC_ARRAY_NAME(lstsq),
         lstsq_types
+    },
+    {
+        "qr",
+        "(m,m),->(m,k),(k,m)",
```

Comment:
`(m, m)` - input dimension, `(m, k)` for Q and `(k, m)` for R. `k = min(m, n)`.

Reply:
I don't think you're allowed to have dimensions that appear only on the rhs.

Also, shouldn't this be `(m, n)` not `(m, m)`?

Reply:
Unfortunately, there are four variants of `qr`:

https://github.com/numpy/numpy/blob/fb215c76967739268de71aa4bda55dd1b062bc2e/numpy/linalg/linalg.py#L788-L791

You'll need a definition for each one.

Reply:
I think you need two definitions each for reduced and raw. For instance, reduced will have two functions: reduced_n with `(m,n) -> (m,n),(n,n)` and reduced_m  `(m,n)->(m,m),(m,n)` (you also have excess commas?) - see lstsq_m and lstsq_n for examples. Then the python code chooses between them 

https://github.com/numpy/numpy/blob/6790873334b143117f4e8d1f515def8c7fdeb9fb/numpy/linalg/linalg.py#L2296-L2300

Reply:
.. assuming you allow non-square matrices

Reply:
> Also, shouldn't this be (m, n) not (m, m)?

Yeah. Actually, I thought that QR decomposition is defined for square matrices so I kept `(m, m)`. However, since, `numpy` allows non-square matrices as well, so we should change the input dimension to `(m, n)`.

> You'll need a definition for each one.

Ah, I see. Thanks for the hint. I will make the changes.